docs: Refactor bash module code guidelines (#175)
blue-build/modules: README.md

This repository contains the default official module repository for [BlueBuild](https://blue-build.org/). See [./modules/](./modules/) for the module source code. See the [website](https://blue-build.org/reference/module/) for module documentation. See [How to make a custom module](https://blue-build.org/how-to/making-modules/) and [Contributing](https://blue-build.org/learn/contributing/) for help with making custom modules and contributing.

## Style & code guidelines for official bash modules

These are general guidelines for writing official bash modules and their documentation to follow in order to keep a consistent style. Not all of these are to be mindlessly followed, especially the ones about grammar and writing style, but it's good to keep these in mind if you intend to contribute back upstream, so that your module doesn't feel out of place.

> ### Terms

Reviewer: This title feels off to me–maybe "Glossarry" or "Definitions" or something would be better?

Author: "Glossary" is what I choose now.

> **Module-maintainer:** Maintainer of the BlueBuild bash module, to whom is this guideline targeted.  

Reviewer: ```suggestion
**Module-maintainer:** Maintainer of a BlueBuild bash module. The intended audience of this section of the documentation.
```

> **Image-maintainer:** Maintainer of the custom image which uses BlueBuild template.  

Reviewer: ```suggestion
**Image-maintainer:** Maintainer of a custom image that uses BlueBuild.
```
Startingpoint was a template. BlueBuild is mostly not.

> **Local-user:** User which uses a custom image, which utilizes BlueBuild bash modules. Image-maintainer can be a local-user too at the same time.

Reviewer: ```suggestion
**Local-user:** User of a custom image using the BlueBuild bash module. The image-maintainer is usually a local-user too.
```

### Code Rules

- Echo what you're doing on each step and on errors to help debugging.

> - Implement error-checks for scenarios where image-maintainer can make a mistake.

Reviewer: ```suggestion
- Implement error-checks for scenarios where the image-maintainer might misconfigure the module.
```

- Use `snake_case` for functions and variables changed by the code.
- Use `SCREAMING_SNAKE_CASE` for variables that are set once and stay unchanged.

> - Use `"${snake_case}"` to ensure that variables are properly parsed for strings.

Reviewer: When should this be used? This could be rewritten to be in the form of "Use `"${variable_name}"` when xyz to ensure that variables are properly parsed as strings."

Author: I rewrote this, tell me if it's good enough, or it needs to be even better

I also added additional rule below, check that too (in separate review thread if fixes are needed).

[`5e37821` (#175)](https://github.com/blue-build/modules/pull/175/commits/5e378219839937b7759dc0aba77642cdf709663a)

Reviewer: LGTM

> - Use `set -euo pipefail` at the start of the script, to ensure that module will fail the image build if error is caught.
> - Use `set -euxo pipefail` at the start of the script for debugging, as each executed command is printed. Only use it for testing & not for final release.

Reviewer: Conflicting instructions. The second one could be rewritten to be a child of the previous item in the list like so:
```suggestion
- Use `set -euo pipefail` at the start of the script, to ensure that module will fail the image build if error is caught.
     -  You can also use `set -euxo pipefail` during debuggiong, so that each executed command is printed. This should not be used in a published module. 
```

For the short module description (`shortdesc:`), the following guidelines apply:
- The description should start with a phrase like "The glorb module reticulates splines" or "The tree module can be used to plant trees".

### Module Config

> **Module config** is used to allow local-users to see & change the behavior of the module on booted system, in order to improve local-user experience.

Reviewer 2: Is this separate from the config passed in via a recipe?

Author: Yes. Maybe it can be rephrased better to not confuse recipe.yml.

Module just passes recipe.yml values to system config.

Otherwise recipe.yml is a separate thing.

Author: I use the term "Post-install module config" now, to not confuse users about "recipe.yml". If better term can be used, feel free to let me know.

- **module performs it's functions on booted system**  
 Modules which are fully utilized in build-time don't need configuration options, as those are already located in `recipe.yml`.
- **module config can be implemented without affecting reliability of the system**  
 Module maintainer needs to carefully select which type of module to implement based on condition above. If a module compromises system reliability when used on booted system, making the module build-time based should be considered. Examples of this are `rpm-ostree` & `akmods` modules, which are better utilized as build-time modules.
- **module can have additional useful options for configuring**  
Which can improve user experience.
- **module can strongly collide with user's usage pattern with it's default behavior**  
Example: `default-flatpaks` module can remove a flatpak app, which local-user used daily.

#### Config Format

In order to keep config files easy to read & reliable to parse, standardized `.yml` markup format is used.  
[`yq`](https://github.com/mikefarah/yq) tool is used to process `.yml` configs in order to reach the desired goal.

#### Config Directory Structure

> **System config:**  
> `/usr/share/bluebuild/module-name/module-config.yml`
> 
> **Local-user config:**  
> `/usr/etc/bluebuild/module-name/module-config.yml` 
> 
> **System config** is a module config which is derived from `recipe.yml` module entry. It is placed in this read-only directory location in order to avoid local-users touching that file.
> 
> **Local-user config** is a module config which is derived from local-user config template. It is placed in `/usr/etc`, which is then automatically copied to `/etc`, which is writable to local-users. `/usr/etc` local-user config can be used to reset module config that is done in `/etc`.

Reviewer 2: I think it would make more sense to group them more appropriately.

```suggestion
**System config:**  
`/usr/share/bluebuild/module-name/module-config.yml`

**System config** is a module config which is derived from `recipe.yml` module entry. It is placed in this read-only directory location in order to avoid local-users touching that file.

**Local-user config:**  
`/usr/etc/bluebuild/module-name/module-config.yml` 

**Local-user config** is a module config which is derived from local-user config template. It is placed in `/usr/etc`, which is then automatically copied to `/etc`, which is writable to local-users. `/usr/etc` local-user config can be used to reset module config that is done in `/etc`.
```

#### Config Example

> System config:

Reviewer 2: Maybe add the example path to the config too.

Author: I already added?

If you mean about adding `recipe.yml` config example, I'm not sure on doing that or how to, as I want to avoid confusion between `recipe.yml` & post-install config.

Author: Oh, I see what you mean. Nvm

For the short module description (`shortdesc:`), the following guidelines apply:
- The description should start with a phrase like "The glorb module reticulates splines" or "The tree module can be used to plant trees".

> ### Local Module Config

Reviewer: This section should also:
- Make the purpose of the configuration as storing data for the module to use when running post-boot clearer.
- Provide some tips and example pieces of code for the implementation.

Author: 1. I think that this makes it clearer
[`37c6433` (#175)](https://github.com/blue-build/modules/pull/175/commits/37c6433320c1af378b5e3a989d0537e67a1a7c0f)
2. I will think about what is the standardized piece of code to present, which would work for every module of this type. I'm open for suggestions.

> ### Post-install Module Config

Reviewer: Should this be called "Local module configuration" instead, as the "local" term is already used?

- **post-install module config can be implemented without affecting reliability of the system**  
 Module-maintainer needs to carefully select which type of module to implement based on condition above. If a module compromises system reliability when used on booted system, making the module build-time based should be considered. Examples of this are `rpm-ostree` & `akmods` modules, which are better utilized as build-time modules.
- **module can have additional useful options for configuring**  
Which can improve local-user experience.
- **module can strongly collide with local-user's usage pattern with it's default behavior**  
Example: `default-flatpaks` module can remove a flatpak app, which local-user used daily.

#### Config Format

In order to keep config files easy to read & reliable to parse, standardized `.yml` markup format is used.  
[`yq`](https://github.com/mikefarah/yq) tool is used to process `.yml` configs in order to reach the desired goal.

#### Config Directory Structure

**System config:**  
`/usr/share/bluebuild/module-name/config.yml`

> **System config** is a post-install module config which is derived from `recipe.yml` module entry. It is placed in this read-only directory location in order to avoid local-users writing to it. System config is purely intended to inform local-users about which modifications are done in `recipe.yml`, so they can potentially proceed with modifications on their own.

Reviewer: Should make the terms clearer here. IMW this is not a post-install config, as it can't be modified, and has the exact same module configuration as the recipe. It is also unclear from the whether this file should include the serialized-to-yaml module config from the recipe directly, or if it should be the data actually used by the module post-boot.

Author: "Post-install" thing is out.

It depends.
Some image-maintainers can include the module, but not configure it with options at all as they don't need it, to satisfy local-users who want it. `initramfs-etc` type of module is a good example for this.
So system config can be empty or with the utilized config options.

Reviewer: I'm not sure whether the usecase of "Including a module but not doing anythign with it incase someone might want it" makes sense. One could always make a new custom image for that module, but most likely there'd be a way to do the same thing locally without the module, right?

Author: I reworded this a bit, tell me if something else needs to be said

Author: This is related to the last thread that you opened in 1st point, so I will resolve this.

**Local-user config:**  
`/usr/etc/bluebuild/module-name/config.yml`

**Local-user config** is a post-install module config which is derived from local-user config template. It is placed in `/usr/etc`, which is then automatically copied to `/etc`, which is writable to local-users. `/usr/etc` local-user config can be used to reset module config that is done in `/etc`.

#### Config Example

System config (`/usr/share/bluebuild/default-flatpaks/config.yml`):

```yaml
vendor: BlueBuild
module: default-flatpaks
description: System config file for `default-flatpaks` BlueBuild module, which is used to install + remove flatpak apps or modify flatpak repos.

> instructions: Read this system config in order to know what is currently configured by the system & what to potentially modify in local-user config (/etc/bluebuild/default-flatpaks/config.yml).

Reviewer: What's the point of this section? Just to inform the reader? Wouldn't comments in the YAML file be a more standard way to do that?

Author: I agree that it's more standardized to use # for comments, but it looks clearer to me like this.
I will try to implement comments that will look clear at the same time.

Reviewer: Yeah, with this it seems that these properties would be used for something. The `vendor` and `module` keys are already in the file path too, but idk...

Author: I just added # as comments for those entries & "# Information about the config file" above.

Tell me if you have a better suggestion.

[`85e5bd8` (#175)](https://github.com/blue-build/modules/pull/175/commits/85e5bd82a946741450454a7a56b40666609e5b43)

Reviewer: Looks great!
